datanommer.models/datanommer/models/__init__.py

```diff
@@ -128,7 +130,16 @@ def add(envelope):
     obj.msg = message['msg']
     obj.headers = headers
 
-    session.add(obj)
+    # Construct an INSERT ... SELECT ... which will do nothing if an
+    # entry with a matching msg_id already exists. This can happen in the
+    # event of message redelivery.
+    cols = [c for c in Message.__table__.c if c.name != 'id']
+    vals = [literal(getattr(obj, c.name)) for c in cols]
+    not_exists = ~exists(select([Message.__table__.c.i]).\
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```

Comment:
I would definitely not prefer raw SQL 😄.

In my opinion, it's fine to trigger a constraint violation in the database to determine that the message already exists. After all, that's one of the great features that a database offers. Exceptions are also a great way for this to be communicated and SQLAlchemy's ORM makes good use of exceptions. Bypassing the ORM layer is, to me, only acceptable in special cases where there are huge performance benefits that outweigh the loss of readability and the "surprise" factor.

> However, if there are two message consumers populating the same database, then user/package creation will race between those two consumers. 

This seems odd to me. If there are two workers, why are they both handling the same message? Is it just an exceptional case where a message might be re-transmitted? If so, it seems like performance shouldn't matter much. If not, what's the point of having two workers?